edge-modules/TwinTester/TwinOperator.cs

```diff
@@ -0,0 +1,354 @@
+        private AnalyzerClient analyzerClient;
+        private Storage storage;
+        private TwinState twinState;
+
+        public TwinOperator(RegistryManager registryManager, ModuleClient moduleClient, AnalyzerClient analyzerClient, Storage storage)
+        {
+            this.registryManager = registryManager;
+            this.moduleClient = moduleClient;
+            this.analyzerClient = analyzerClient;
+            this.storage = storage;
+            this.twinState = this.InitializeModuleTwin().Result;
+            this.moduleClient.SetDesiredPropertyUpdateCallbackAsync(this.OnDesiredPropertyUpdateAsync, storage);
+        }
+
+        private int GetNewPropertyCounter(Dictionary<string, DateTime> properties)
+        {
+            int maxPropertyId = -1;
+            foreach (KeyValuePair<string, DateTime> propertyUpdate in properties)
+            {
+                maxPropertyId = Math.Max(int.Parse(propertyUpdate.Key), maxPropertyId);
+            }
+
+            return maxPropertyId + 1;
+        }
+
+        private TwinCollection GetReportedPropertiesResetTwin(Twin originalTwin)
+        {
+            TwinCollection eraseReportedProperties = new TwinCollection();
+            foreach (dynamic twinUpdate in originalTwin.Properties.Reported)
+            {
+                KeyValuePair<string, object> pair = (KeyValuePair<string, object>)twinUpdate;
+                eraseReportedProperties[pair.Key] = null; // erase twin property by assigning null
+            }
+
+            return eraseReportedProperties;
+        }
+
+        private async Task<TwinState> InitializeModuleTwin()
+        {
+            while (true)
+            {
+                try
+                {
+                    TwinState initializedState;
+                    Twin twin = await this.registryManager.GetTwinAsync(Settings.Current.DeviceId, Settings.Current.ModuleId);
+                    int storageCount = (await this.storage.GetAllDesiredPropertiesReceived()).Count + (await this.storage.GetAllDesiredPropertiesUpdated()).Count + (await this.storage.GetAllReportedPropertiesUpdated()).Count;
+                    if (storageCount == 0)
+                    {
+                        Logger.LogInformation("No existing storage detected. Initializing new module twin for fresh run.");
+
+                        // reset desired properties
+                        Twin desiredPropertyResetTwin = await this.registryManager.ReplaceTwinAsync(Settings.Current.DeviceId, Settings.Current.ModuleId, new Twin(), twin.ETag);
+
+                        // reset reported properties
+                        TwinCollection eraseReportedProperties = this.GetReportedPropertiesResetTwin(desiredPropertyResetTwin);
+                        await this.moduleClient.UpdateReportedPropertiesAsync(eraseReportedProperties);
+
+                        await Task.Delay(1000 * 10); // give ample time for reported properties reset to reach cloud
```

Comment:
pls use TimeSpan

Reply:
Do we need this? If the update succeeds, it should be updated, right?

Reply:
If the reported property update succeeds, it needs a bit of time before we can perform a reliable getTwin() to the cloud, or else we will be getting an old twin.